packages/core/src/web/app/actions/beambox/beambox-preference.ts

```diff
@@ -1,73 +1,205 @@
+  'default-autofocus': false,
+  'default-borderless': false,
+  'default-diode': false,
+  'default-laser-module': LayerModule.LASER_20W_DIODE,
+  'diode-one-way-engraving': true,
   diode_offset_x: constant.diode.defaultOffsetX,
   diode_offset_y: constant.diode.defaultOffsetY,
-  engrave_dpi: 'medium', // low, medium, high
+  'enable-custom-backlash': false,
+  'enable-custom-preview-height': false,
+  'enable-job-origin': false,
+  enable_mask: false,
+  engrave_dpi: 'medium',
+  'extend-rotary-workarea': false,
+  fast_gradient: true,
+  'font-convert': '2.0',
+  'font-substitute': true,
+  'frame-before-start': false,
   guide_x0: 0,
   guide_y0: 0,
+  image_downsampling: true,
+  'job-origin': 1,
+  'keep-preview-result': false,
   low_power: 10,
   model: 'fbb1b',
+  'module-offsets': moduleOffsets,
   mouse_input_device: window.os === 'MacOS' ? 'TOUCHPAD' : 'MOUSE',
-  rotary_mode: 0,
+  'multipass-compensation': true,
+  'one-way-printing': true,
+  padding_accel: 5000,
+  padding_accel_diode: 4500,
+  'pass-through': false,
+  'path-engine': 'fluxghost',
+  precut_x: 0,
+  precut_y: 0,
+  preview_movement_speed_level: PreviewSpeedLevel.SLOW,
+  'print-advanced-mode': false,
+  'promark-start-button': false,
+  'reverse-engraving': false,
+  'rotary-chuck-obj-d': CHUCK_ROTARY_DIAMETER,
+  'rotary-mirror': false,
+  'rotary-type': RotaryType.Roller,
+  'rotary-y': null,
+  rotary_mode: false,
+  rotary_y_coord: 5,
   should_remind_calibrate_camera: true,
   show_grids: true,
   show_guides: false,
+  show_rulers: false,
+  simplify_clipper_path: false,
   use_layer_color: true,
+  vector_speed_constraint: true,
+  workarea: 'fbb1b',
+  zoom_with_window: false,
 };
 
 const eventEmitter = eventEmitterFactory.createEventEmitter('beambox-preference');
 
-class BeamboxPreference {
+const objectKeys = ['customized-dimension', 'module-offsets'] as const;
+
+class BeamboxPreferenceClass {
   constructor() {
     // set default preference if key or even beambox-preference doesn't exist
-    let pref: any = storage.get('beambox-preference');
-
-    pref = pref === '' ? {} : pref;
-    console.log(pref);
-
-    const fullPref = Object.assign(DEFAULT_PREFERENCE, pref);
-
-    storage.set('beambox-preference', fullPref);
+    const preference: BeamboxPreference = storage.get('beambox-preference') || DEFAULT_PREFERENCE;
+
+    // to migrate preference of old version
+    for (const key in DEFAULT_PREFERENCE) {
+      if (!(key in preference)) {
+        // @ts-expect-error key is keyof BeamboxPreference
+        preference[key] = DEFAULT_PREFERENCE[key];
+      }
+
+      if (objectKeys.includes(key)) {
+        for (const subKey in DEFAULT_PREFERENCE[key as (typeof objectKeys)[number]]) {
+          if (!(subKey in preference[key as (typeof objectKeys)[number]])) {
+            // @ts-expect-error key is keyof BeamboxPreference
+            preference[key] = { ...preference[key], [subKey]: DEFAULT_PREFERENCE[key][subKey] };
+          }
+        }
```

Comment:
Feel like this for loop is equal to `preference[key] = { ...DEFAULT_PREFERENCE[key], ...preference[key]  }`